index.bs

```diff
@@ -6014,6 +6061,85 @@ To <dfn data-lt="bidirectionally broadcasting the shapes">bidirectionally broadc
+
+<details open algorithm="cast">
+<summary>
+To <dfn>cast</dfn> a number |x| to a given {{MLOperandDataType}} |dataType|, perform the following steps. They return a number.
+</summary>
+
+1. Switch on |dataType|:
+    <dl class=switch>
+        : {{MLOperandDataType/"float32"}}
+        :: Return [=ConvertToFloat=](|x|, 32).
+        : {{MLOperandDataType/"float16"}}
+        :: Return [=ConvertToFloat=](|x|, 16).
+        : {{MLOperandDataType/"int64"}}
+        :: Return [=ConvertToInt=](|x|, 64, "signed").
+        : {{MLOperandDataType/"uint64"}}
+        :: Return [=ConvertToInt=](|x|, 64, "unsigned").
+        : {{MLOperandDataType/"int32"}}
+        :: Return [=ConvertToInt=](|x|, 32, "signed").
+        : {{MLOperandDataType/"uint32"}}
+        :: Return [=ConvertToInt=](|x|, 32, "signed").
+        : {{MLOperandDataType/"int8"}}
+        :: Return [=ConvertToInt=](|x|, 8, "signed").
+        : {{MLOperandDataType/"uint8"}}
+        :: Return [=ConvertToInt=](|x|, 8, "unsigned").
+    </dl>
+
+NOTE: The input to [=cast=] is an abstract number with unlimited range and precision, including the special values Infinity, -Infinity and NaN. The output is also an abstract number, but exactly representable as the specified type.
+
+<div algorithm>
+The <dfn>ConvertToFloat</dfn>(|x|, |bitLength|) steps are:
+
+1. If |x| is NaN, then return NaN.
+1. Switch on |bitLength|:
+    <dl class=switch>
+        : 32
+        ::
+            1. Let |upperBound| be 2<sup>128</sup>.
+            1. Let |lowerBound| be -2<sup>128</sup>.
+            1. Let |S| be the set of [[IEEE-754-2019]] binary32 floating point values except -0, but with the special values |upperBound| and |lowerBound| added.
+        : 16
+        ::
+            1. Let |upperBound| be 2<sup>16</sup>.
+            1. Let |lowerBound| be -2<sup>16</sup>.
+            1. Let |S| be the set of [[IEEE-754-2019]] binary16 floating point values except -0, but with the special values |upperBound| and |lowerBound| added.
+    </dl>
+1. Let |y| be the number in |S| that is closest to |x|, selecting the number with an even significand if there are two [=equally close values=]. The two special values |lowerBound| and |upperBound| are considered to have even significands for this purpose.
+1. If |y| is |upperBound|, then return +Infinity.
+1. If |y| is |lowerBound|, then return -Infinity.
+1. If |y| is +0 and |x| is negative, return -0.
+1. Return |y|.
+
+NOTE: This is based on a definition in [[WEBIDL]], but extended to cover 16-bit floating point values.
+
+</div>
+
+<div algorithm>
+The <dfn>ConvertToInt</dfn>(|x|, |bitLength|, |signedness|) steps are:
+
+1. If |signedness| is "unsigned", then:
+    1. Let |lowerBound| be 0.
+    1. Let |upperBound| be 2<sup>|bitLength|</sup> - 1.
+1. Otherwise:
+    1. Let |lowerBound| be -2<sup>|bitLength| - 1</sup>.
```

Comment:
> -2<sup>|bitLength| - 1</sup>

Huh, so I see some other mathy programs following this surprising behavior (surprising to me anyway), but this just seems so fundamentally wrong that some operator like multiplication or exponentiation should bind more tightly to a number than a number should bind to that number's *own sign*. Like, "-2" is an atomic entity (the negative sign is not some negating operator, nor is it subtraction in PEMDAS, but rather part of the identity of the numeric literal itself), and thus `-2 ^ 16` *should* mean `pow(-2, 16)`, and not `-pow(2, 16)`. Well, I don't have the energy to fight the historical faux pas of potentially multiple math books. So, mind just clarifying the order operations with some parens? :b

```suggestion
    1. Let |lowerBound| be -(2<sup>|bitLength| - 1</sup>).
```